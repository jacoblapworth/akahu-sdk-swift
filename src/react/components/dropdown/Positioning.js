import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import verge from 'verge';
import throttle from 'lodash.throttle';
import Portal from 'react-portal';
import cn from 'classnames';
import breakpoints from './private/breakpoints';


/**
 * Tests the height and width of the given rectangle and returns true if the size is greater than 0x0
 *
 * @private
 * @param {DOMRect} baseRect
 * @return {boolean}
 */
const isBaseRendered = baseRect => baseRect && baseRect.height > 0 && baseRect.width > 0;

/**
* Tests the viewport against a narrow width.
*
* @private
* @return {boolean}
*/
const isNarrowViewport = () => verge.viewportW() < breakpoints.narrow;

/**
* Calculates the space below the trigger
*
* @private
* @param {DOMRect} triggerRect
* @return {Number}
*/
const calcSpaceBelow = triggerRect => triggerRect && verge.viewportH() - triggerRect.top - triggerRect.height;

/**
* Calulates the space above the trigger
*
* @private
* @param {DOMRect} triggerRect
* @return {Number}
*/
const calcSpaceAbove = triggerRect => triggerRect.top;

/**
* Tests if the base can be rendered under the trigger at all.
*
* @private
* @param {DOMRect} triggerRect
* @param {DOMRect} popupRect
* @param {Number} gutter
* @return {Boolean}
*/
const canAlignUnderTrigger = (triggerRect, popupRect, gutter) => popupRect.height + gutter < calcSpaceBelow(triggerRect);

/**
* Returns the top scroll amount, supported across mutliple browsers
*
* @private
* @return {Number}
*/
const scrollTopAmount = () => window.pageYOffset || document.documentElement.scrollTop || document.body.scrollTop || 0;

/**
* Returns the left scroll amount, supported across mutliple browsers
*
* @private
* @return {Number}
*/
const scrollLeftAmount = () => window.pageXOffset || document.documentElement.scrollLeft || document.body.scrollLeft || 0

/**
* Tests if the bottom space is greater than the top around the trigger.
*
* @private
* @param {DOMRect} triggerRect
* @return {Boolean}
*/
const isBottomBiggerThanTop = triggerRect => {
	const spaceAbove = calcSpaceAbove(triggerRect);
	const spaceBelow = calcSpaceBelow(triggerRect);

	return spaceBelow > spaceAbove;
}

/**
 * Listeners to attach to window
 *
 * @private
 * @param {Positioning} popup
 */
function attachListeners(popup) {
	window.addEventListener('resize', popup.positionComponent);
	window.addEventListener('scroll', popup.positionComponent);
	if (popup.props.setMaxHeight) {
		window.addEventListener('resize', popup.calculateMaxHeight)
	}
}

/**
 * Listeners to detach from window
 *
 * @private
 * @paramn {Positioning} popup
 */
function detachListeners(popup) {
	window.removeEventListener('resize', popup.positionComponent);
	window.removeEventListener('scroll', popup.positionComponent);
	window.removeEventListener('resize', popup.calculateMaxHeight)
}

/**
 * @private
 * Calculates the base position and adds it to state, including a flag so we know the calculations have been done.
 *
 * @param {Object} popupRect
 * @param {Object} triggerRect
 * @param {Positioning} popup
 */
function alignBaseWithTrigger(popupRect, triggerRect, popup) {
	const { gutter } = popup.props;

	// If there's not enough space below the trigger and there's more space above the trigger than below, render above
	const alignTop = !isBottomBiggerThanTop(triggerRect) && !canAlignUnderTrigger(triggerRect, popupRect, gutter);
	const alignRight = verge.viewportW() - popupRect.width < triggerRect.left + gutter;

	const positionBelowTrigger = triggerRect.top + triggerRect.height;
	const positionAboveTrigger = triggerRect.top - popupRect.height;
	const triggerOffsetTop = (alignTop ? positionAboveTrigger: positionBelowTrigger) + scrollTopAmount();

	const positionRightOfTrigger = triggerRect.left + triggerRect.width - popupRect.width;
	const positionLeftOfTrigger = triggerRect.left;
	const triggerOffsetLeft = (alignRight ? positionRightOfTrigger : positionLeftOfTrigger) + scrollLeftAmount();

	popup.setState({
		left: triggerOffsetLeft,
		top: triggerOffsetTop,
		alignTop: alignTop
	});
}

/**
* Aligns the popup to the very top left of the viewport, commonly useful for narrow viewport widths.
*
* @private
* @param {Positioning} popup
*/
function alignTopLeft(popup) {
		popup.setState({
			left: 0,
			top: 0
		});
}

/**
 * In order to ensure that measurements are taken after all children are rendered, we need to do a setTimeout inside
 * of the componentDidUpdate method.  This is the callback for that and will correctly position the element with a
 * computed maxHeight, if desired.
 *
 * @private
 * @param {Positioning} popup
 */
function positionOnShow(popup) {
	// Safety check due to slim chance of unmount during setTimeout duration
	if (popup.positionEl && document.body.contains(popup.positionEl)) {
		popup.positionComponent();
		if (popup.props.setMaxHeight) {
			popup.calculateMaxHeight();
		}
		// Tell the render method it's OK to render without "visibility: hidden"
		popup.setState({
			positioned: true
		});
	}
}

/**
 * Create a copy of a clean, initial state of this component.  Used when setting state during construction and clearing
 * state when hiding the component.
 *
 * @private
 * @returns {{alignRight: boolean, alignTop: boolean, maxHeight: string, positioned: boolean}}
 */
function getDefaultState() {
	return {
		top: 0,
		left: 0,
		alignTop: false,
		maxHeight: 0,
		positioned: false
	};
}

/**
 * While we're calculating where the popup should go, it needs to have a set of styles which makes it invisible with
 * proper dimensions without affecting the layout of the page in general.  These styles will take care of that.
 *
 * @private
 * @param {Positioning} popup
 * @returns {Object|null}
 */
function getPositionCalculationStyles(popup) {
	if (!popup.props.renderHidden && !popup.state.positioned) {
		return {
			visibility: 'hidden',
			position: 'fixed',
			top: 0,
			left: 0,
			maxHeight: '99%',
			maxWidth: '99%'
		};
	}
	return null;
}

class Positioning extends PureComponent {
	constructor() {
		super();

		const popup = this;

		popup.state = getDefaultState();

		popup.positionComponent = throttle(popup.positionComponent.bind(popup), 100, { trailing: true });
		popup.calculateMaxHeight = throttle(popup.calculateMaxHeight.bind(popup), 100, { trailing: true });
	}

	componentWillUnmount() {
		// Safety cleanup
		detachListeners(this);
	}

	/**
	 * Calculate positioning of the popup if the trigger is rendered.
	 */
	positionComponent() {
		const popup = this;
		const { parentRef } =  popup.props;

		if (parentRef) {
			const triggerDOM = parentRef.firstChild;
			const baseRect =  popup.positionEl && popup.positionEl.firstChild.getBoundingClientRect();

			if (isBaseRendered(baseRect) && verge.inViewport(triggerDOM)) {
				isNarrowViewport() ? alignTopLeft(popup) : alignBaseWithTrigger(baseRect, triggerDOM.getBoundingClientRect(), popup);
			}
		}
	}

	/**
	 * Given we're rendering in the greatest whitespace, we need to work out if a maxHeight should be added to the popup
	 * in order for the full popup to show its content and scroll.
	 */
	calculateMaxHeight() {
		const popup = this;
		const { gutter, parentRef } = popup.props;
		const triggerDOM = parentRef.firstChild;

		if (verge.inViewport(triggerDOM)) {
			const triggerRect = triggerDOM.getBoundingClientRect();

			const spaceAbove = calcSpaceAbove(triggerRect);
			const spaceBelow = calcSpaceBelow(triggerRect);

			popup.setState({
				maxHeight: isNarrowViewport() ? verge.viewportH() : Math.max(spaceAbove, spaceBelow) - gutter
			});
		}
	}

	componentDidUpdate(prevProps) {
		const {props, state} = this;

		// If the popup is going from hidden to visible but hasn't been positioned yet, the render method will ensure
		// that everything is rendered with "visibility: hidden".  Wait a bit to make sure all children also render,
		// then measure things and position the popup correctly on the screen.
		if (!props.renderHidden && !state.positioned) {
			setTimeout(positionOnShow, 100, this);
		}

		// In order to ensure that the event listeners are properly attached based on a potentially changing
		// setMaxHeight prop when visible and detached when hidden, we'll need to unconditionally detach then reattach
		// if visible here.
		detachListeners(this);
		if (!props.renderHidden) {
			attachListeners(this);
		}

		// If we're hiding the popup, reset the state to defaults so that the next show event will properly reposition
		// everything.
		if (!prevProps.renderHidden && props.renderHidden) {
			this.setState(getDefaultState());
		}
	}

	/**
	 * @public
	 * Uses internal state to work out inline styles and returns them.
	 *
	 * @return {{position: string}}
	 */
	getStyles() {
		const { maxHeight, left, top } = this.state;

		return {
			maxHeight,
			left,
			top
		};
	}

	render() {
		const popup = this;
		const { children, renderHidden } = popup.props;
		const { positioned } = popup.state;
		const positioningStyles = getPositionCalculationStyles(popup);
		const clonedChildren = renderHidden || !positioned ? children : React.cloneElement(children, {
			className : cn( children.props.className, { 'dropdown-positionabove' : popup.state.alignTop } ),
			style : popup.getStyles()
		});

		return (
			<Portal isOpened={!renderHidden} onOpen={popup.positionComponent}>
				<div style={positioningStyles} ref={portal => popup.positionEl = portal}>
					{clonedChildren}
				</div>
			</Portal>
		);
	}
}

Positioning.propTypes = {
	className: PropTypes.string,
	children: PropTypes.node,
	/** @property {renderHidden} true when the component is rendered but not displayed */
	renderHidden: PropTypes.bool,
	/** @property {parentRef} A DOM object of the parent node. */
	parentRef: PropTypes.object,
	/** @property {gutter} A buffer value added to measure between the edge of the viewport and the component before flipping its position. */
	gutter: PropTypes.number,
	/** @property {setMaxHeight} A max height will mean an overflowed popup will scroll for the user rather than render outside of the viewport. True by default. */
	setMaxHeight: PropTypes.bool
};

Positioning.defaultProps = {
	gutter: 20,
	setMaxHeight: true
};

export default Positioning;
