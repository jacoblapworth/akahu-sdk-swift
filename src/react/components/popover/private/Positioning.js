import PropTypes from 'prop-types';
import React from 'react';
import { Portal } from 'react-portal';

import shallowCompare from '../../../helpers/shallowCompare';
import PositioningHelper from './helpers/positioning';
import doAsync from '../../helpers/doAsync';

export default class Positioning extends React.Component {
  ref = React.createRef();

  updatePositionTimeoutId;

  windowHasResizeListener = false;

  state = {};

  componentDidMount() {
    this.updatePosition(true);

    if (document.readyState !== 'complete') {
      document.addEventListener('DOMContentLoaded', this.updatePosition);
    }

    if (window) {
      window.addEventListener('resize', this.updatePosition);
      this.windowHasResizeListener = true;
    }
  }

  componentWillUnmount() {
    window && window.removeEventListener('resize', this.updatePosition);
  }

  componentDidUpdate() {
    this.updatePosition(true);
    if (!this.windowHasResizeListener && window) {
      window.addEventListener('resize', this.updatePosition);
      this.windowHasResizeListener = true;
    }
  }

  render() {
    const { children } = this.props;
    const { isFullWidth, location, style } = this.state;
    return (
      <Portal>
        <div className="xui-container">
          <div ref={this.ref} style={{ position: 'absolute', ...style }}>
            {children(location, isFullWidth)}
          </div>
        </div>
      </Portal>
    );
  }

  /* eslint-disable no-param-reassign */
  getContentRect = contentElement => {
    const previousStyle = { left: contentElement.style.left, top: contentElement.style.top };
    /**
     * We place the contentElement at the top left of the page before retrieving its measurements so
     * that the size is not influenced by the edges of the page.
     */
    contentElement.style.left = '0';
    contentElement.style.top = '0';

    const contentRect = contentElement.getBoundingClientRect();

    contentElement.style.left = previousStyle.left;
    contentElement.style.top = previousStyle.top;

    return contentRect;
  };
  /* eslint-enable no-param-reassign */

  updatePosition = async updateSynchronously => {
    const doSync = callback => new Promise(resolve => resolve(callback()));
    const execute = updateSynchronously ? doSync : doAsync;

    const { pageGutter, preferredLocation, triggerGap } = this.props;

    if (this.updatePositionTimeoutId) {
      clearTimeout(this.updatePositionTimeoutId);
    }

    if (!this.ref.current || !this.props.triggerRef.current) {
      return;
    }

    const contentRect = await execute(
      () => this.ref.current && this.getContentRect(this.ref.current),
    );
    const triggerRect = await execute(() => this.props.triggerRef.current?.getBoundingClientRect());

    if (triggerRect) {
      const positionHelper = await execute(
        () =>
          new PositioningHelper(
            preferredLocation,
            contentRect,
            triggerRect,
            triggerGap,
            pageGutter,
          ),
      );

      const location = await execute(() => positionHelper.getLocation());
      const style = await execute(() => positionHelper.getPositionStyle());
      const isFullWidth = await execute(() => positionHelper.isFullWidth());
      // This function is called often so we only call `setState` if something has changed.

      if (location !== this.state.location || !shallowCompare(style, this.state.style)) {
        this.setState({
          isFullWidth,
          location,
          style,
        });
      }
    }

    /**
     * Running `updatePosition` regularly accommodates the following situations:
     * - The content resizes (could be handled with a ResizeObserver)
     * - The trigger resizes (could be handled with a ResizeObserver)
     * - The trigger changes position (currently we have no way of observing this reliably)
     */
    this.updatePositionTimeoutId = setTimeout(this.updatePosition, 100);
  };
}

Positioning.propTypes = {
  /**
   * This component requires `children` to be a function. The function receives 2 arguments,
   * `location` and `isFullWidth`, and returns a React node.
   *
   * The `location` is where the content ended up being positioned, this will be the
   * `preferredLocation` or the next available location.
   *
   * The `isFullWidth` flag is `true` when the width of the content is narrower than the body when
   * taking the `pageGutter` into consideration. This is useful for switching to a full-width style
   * when there is not enough room to position the content correctly.
   */
  children: PropTypes.func.isRequired,
  /**
   * How much space should be left between the edge of the page and the content to be positioned.
   */
  pageGutter: PropTypes.number,
  /**
   * The preferred location for the content. If the preferred location is not available the opposite
   * side of the trigger will be checked, followed by the right and bottom sides, and finally the
   * left and top sides.
   */
  preferredLocation: PropTypes.oneOf(['bottom', 'left', 'right', 'top']),
  /**
   * How much space should be left between the trigger and the content to be positioned.
   */
  triggerGap: PropTypes.number,
  /**
   * The ref for the HTMLElement the content should be positioned around.
   */
  triggerRef: PropTypes.object,
};
