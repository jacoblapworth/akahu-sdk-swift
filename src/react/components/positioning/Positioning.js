import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import verge from 'verge';
import Portal from 'react-portal';
import debounce from 'lodash.debounce';
import cn from 'classnames';
import {
	isNarrowViewport,
	calcSpaceAbove,
	calcSpaceBelow,
	calcSpaceRight,
	calcSpaceLeft,
	scrollTopAmount,
	scrollLeftAmount,
	isBaseRendered,
	attachListeners,
	detachListeners,
} from './private/dom-helpers';

/**
 * @private
 * Calculates the base position and adds it to state, including a flag so we know the calculations have been done.
 *
 * @param {Object} popupRect
 * @param {Object} triggerRect
 * @param {Positioning} popup
 */
function alignBaseWithTrigger(popupRect, triggerRect, popup) {
	const { viewportGutter, triggerDropdownGap } = popup.props;
	const spaceAboveTrigger = calcSpaceAbove(triggerRect);
	const spaceBelowTrigger = calcSpaceBelow(triggerRect);
	const spaceLeftOfTrigger = calcSpaceLeft(triggerRect);
	const spaceRightOfTrigger = calcSpaceRight(triggerRect);

	const canGoBelow = popupRect.height <= spaceBelowTrigger - viewportGutter - triggerDropdownGap;
	const canAlignLeftEdge = popupRect.width <= spaceRightOfTrigger + triggerRect.width - viewportGutter - triggerDropdownGap;

	const placeBelow = canGoBelow || spaceBelowTrigger >= spaceAboveTrigger;
	const alignLeftEdge = canAlignLeftEdge || spaceRightOfTrigger >= spaceLeftOfTrigger;

	const popupLeftPos = alignLeftEdge
		? Math.max(triggerRect.left, viewportGutter)
		: Math.min(Math.max(triggerRect.right - popupRect.width, viewportGutter), verge.viewportW() - popupRect.width - viewportGutter);

	const translateX = isNarrowViewport()
		? '0px'
		: `${Math.floor(popupLeftPos + scrollLeftAmount())}px`
	const translateY = placeBelow
		? `${triggerRect.height}px`
		: '-100%';
	const translate = `translate(${translateX},${translateY})`;
	// Initially the gap offset here was done through css calc properties in the translate function. Unfortunately
	// this caused issues, as calc is invalid as a parameter of translate within IE11
	const topValue = placeBelow ?
		triggerRect.top + scrollTopAmount() + triggerDropdownGap
		: triggerRect.top + scrollTopAmount() - triggerDropdownGap;

	popup.setState({
		top: topValue,
		alignTop: !placeBelow,
		transform: translate,
		bottom: null,
	});
}

/**
* Aligns the popup to the very bottom left of the viewport, commonly useful for narrow viewport widths.
*
* @private
* @param {Positioning} popup
*/
function alignBottomLeft(popup) {
		popup.setState({
			bottom: 0,
			left: 0,
			maxHeight: null,
			transformY: null,
			top: null,
		});
}

/**
 * In order to ensure that measurements are taken after all children are rendered, we need to do a setTimeout inside
 * of the componentDidUpdate method.  This is the callback for that and will correctly position the element with a
 * computed maxHeight, if desired.
 *
 * @private
 * @param {Positioning} popup
 */
function positionOnShow(popup) {
	// Safety check due to slim chance of unmount during setTimeout duration
	if (popup.positionEl && document.body.contains(popup.positionEl)) {
		popup.positionComponent();
		if (popup.props.shouldRestrictMaxHeight) {
			popup.calculateMaxHeight();
		}
		// Tell the render method it's OK to render without "visibility: hidden"
		popup.setState({
			positioned: true,
		});
	}
}

/**
 * Create a copy of a clean, initial state of this component.  Used when setting state during construction and clearing
 * state when hiding the component.
 *
 * @private
 * @returns {{alignRight: boolean, alignTop: boolean, maxHeight: string, positioned: boolean}}
 */
function getDefaultState() {
	return {
		top: 0,
		transform: null,
		alignTop: false,
		maxHeight: verge.viewportH() * 0.99,
		positioned: false,
	};
}

/**
 * While we're calculating where the popup should go, it needs to have a set of styles which makes it invisible with
 * proper dimensions without affecting the layout of the page in general.  These styles will take care of that.
 *
 * @private
 * @param {Positioning} popup
 * @returns {Object|null}
 */
function getPositionCalculationStyles(popup) {
	if (popup.props.isVisible && !popup.state.positioned) {
		return {
			visibility: 'hidden',
			position: 'fixed',
			top: 0,
			transform: null,
			maxHeight: '99%',
			maxWidth: '99%'
		};
	}
	return null;
}

class Positioning extends PureComponent {
	constructor() {
		super();

		const popup = this;

		popup.state = getDefaultState();

		popup.positionComponent = popup.positionComponent.bind(popup);
		popup.calculateMaxHeight = popup.calculateMaxHeight.bind(popup);
		popup.resizeHandler = debounce(popup.positionComponent, 75, { leading: false, trailing: true });
	}

	componentDidMount() {
		if (this.props.isVisible) {
			attachListeners(this);
		}
	}

	componentDidUpdate(prevProps, prevState) {
		const { props, state } = this;

		// If the popup is going from hidden to visible but hasn't been positioned yet, the render method will ensure
		// that everything is rendered with "visibility: hidden".  Wait a bit to make sure all children also render,
		// then measure things and position the popup correctly on the screen.
		if (props.isVisible && !state.positioned) {
			this._positionTimer = setTimeout(positionOnShow, 50, this);
		}

		if (!prevState.positioned && state.positioned && props.onVisible != null) {
			this._visibleTimer = setTimeout(props.onVisible, 50);
		}

		if (props.isVisible !== prevProps.isVisible) {
			if (!props.isVisible) {
				// If we're hiding the popup, reset the state to defaults so that the next show event will properly
				// reposition everything.
				this.setState(getDefaultState());
				detachListeners(this);

				// In case these haven't fired for some reason, kill them now to prevent errors
				clearTimeout(this._positionTimer);
				clearTimeout(this._visibleTimer);
			} else {
				attachListeners(this);
			}
		}

		if (prevProps.shouldRestrictMaxHeight !== props.shouldRestrictMaxHeight) {
			if (props.shouldRestrictMaxHeight) {
				this.calculateMaxHeight();
			} else {
				this.setState({
					maxHeight: null,
				});
			}
		}
	}

	componentWillUnmount() {
		detachListeners(this);
		clearTimeout(this._positionTimer);
		clearTimeout(this._visibleTimer);
	}

	/**
	 * Calculate positioning of the popup if the trigger is rendered.
	 *
	 * @public
	 */
	positionComponent() {
		const popup = this;
		const { parentRef } = popup.props;

		if (parentRef) {
			const triggerDOM = parentRef.firstChild;
			const baseRect = popup.positionEl && popup.positionEl.firstChild.getBoundingClientRect();

			if (isBaseRendered(baseRect)) {
				if (!popup.props.isNotResponsive && isNarrowViewport()) {
					alignBottomLeft(popup);
				} else {
					alignBaseWithTrigger(baseRect, triggerDOM.getBoundingClientRect(), popup);
				}
			}
		}
	}

	/**
	 * Given we're rendering in the greatest whitespace, we need to work out if a maxHeight should
	 * be added to the popup in order for the full popup to show its content and scroll.
	 *
	 * @public
	 */
	calculateMaxHeight() {
		const popup = this;
		const { viewportGutter, parentRef, triggerDropdownGap } = popup.props;
		const triggerDOM = parentRef.firstChild;

		if (verge.inViewport(triggerDOM)) {
			if (!popup.props.isNotResponsive && isNarrowViewport()) {
				popup.setState({
					maxHeight: verge.viewportH(),
				});
			} else {
				const triggerRect = triggerDOM.getBoundingClientRect();
				const spaceAboveTrigger = calcSpaceAbove(triggerRect);
				const spaceBelowTrigger = calcSpaceBelow(triggerRect);

				popup.setState({
					maxHeight: Math.max(spaceAboveTrigger, spaceBelowTrigger) - viewportGutter - triggerDropdownGap,
				});
			}
		}
	}

	/**
	 * Uses internal state to work out inline styles and returns them.
	 *
	 * @return {{ maxHeight: Number, left: Number, top: Number, transformY: String }}
	 */
	getStyles() {
		const { maxHeight, transform, top, bottom } = this.state;
		const { isTriggerWidthMatched, parentRef, isNotResponsive } = this.props;
		const isMobile = isNarrowViewport() && !isNotResponsive;
		let width = null;
		let maxWidth = null;
		if (isTriggerWidthMatched && !isMobile && parentRef != null && parentRef.firstChild != null) {
			width = parentRef.firstChild.getBoundingClientRect().width;
			maxWidth = 'none';
		}

		return {
			maxHeight: isMobile ? null : maxHeight,
			top,
			width,
			maxWidth,
			bottom,
			transform: isMobile ? '' : transform,
		};
	}

	render() {
		const popup = this;
		const { children, isVisible, qaHook } = popup.props;
		const { positioned } = popup.state;
		const positioningStyles = getPositionCalculationStyles(popup);
		const clonedChildren = !isVisible || !positioned ? children : React.cloneElement(children, {
			className : cn( children.props.className, { 'dropdown-positionabove' : popup.state.alignTop } ),
			style : popup.getStyles(),
		});

		return (
			<Portal isOpened={isVisible} onOpen={popup.positionComponent} >
				<div style={positioningStyles} ref={portal => popup.positionEl = portal} className="xui-container" data-automationid={qaHook}>
					{clonedChildren}
				</div>
			</Portal>
		);
	}
}

Positioning.propTypes = {
	className: PropTypes.string,
	children: PropTypes.node,
	qaHook: PropTypes.string,
	/**true when the component is rendered but not displayed */
	isVisible: PropTypes.bool,
	/**A DOM object of the parent node. */
	parentRef: PropTypes.object,
	/**A buffer value added to measure between the edge of the viewport and the component before flipping its position. */
	viewportGutter: PropTypes.number,
	/**A max height will mean an overflowed popup will scroll for the user rather than render outside of the viewport. True by default. */
	shouldRestrictMaxHeight: PropTypes.bool,
	/** Force the desktop UI, even if the viewport is narrow enough for mobile. */
	isNotResponsive: PropTypes.bool,
	/** The amount of space to put between the trigger and the dropdown */
	triggerDropdownGap: PropTypes.number,
	/** Callback for when the positioned element becomes visible  */
	onVisible: PropTypes.func,
	/** Setting to true will for the dropdown to be as wide as the trigger. */
	isTriggerWidthMatched: PropTypes.bool,
};

Positioning.defaultProps = {
	viewportGutter: 10,
	shouldRestrictMaxHeight: true,
	isNotResponsive: false,
	triggerDropdownGap: 5,
	isTriggerWidthMatched: false,
};

export default Positioning;
