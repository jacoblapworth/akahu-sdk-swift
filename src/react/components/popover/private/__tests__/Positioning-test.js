import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import React from 'react';

import Positioning from '../Positioning';

Enzyme.configure({ adapter: new Adapter() });

describe('<Positioning />', () => {
  it('renders without crashing', () => {
    // Arrange
    const triggerRef = React.createRef();
    const wrapper = shallow(
      <Positioning triggerRef={triggerRef}>{side => <div>{side}</div>}</Positioning>,
    );

    // Assert
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
