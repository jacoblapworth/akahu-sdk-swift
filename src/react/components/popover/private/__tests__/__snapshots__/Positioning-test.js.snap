// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Positioning /> renders without crashing 1`] = `
<Portal>
  <div
    className="xui-container"
  >
    <div
      style={
        Object {
          "position": "absolute",
        }
      }
    >
      <div />
    </div>
  </div>
</Portal>
`;
